src/StringUtils.cpp

```diff
@@ -14,196 +14,373 @@
 #include <numeric>
 #include <sstream>
 
-/**
- * \fn toLowercase(const std::string& str)
- *
- * Converts a string to lowercase.
- *
- * \param str	Source string.
- *
- * \return	Returns the converted string.
- */
-std::string NAS2D::toLowercase(std::string str)
+namespace
 {
-	std::transform(std::begin(str), std::end(str), std::begin(str), [](unsigned char c) noexcept->unsigned char { return static_cast<unsigned char>(::tolower(c)); });
-	return str;
+	template<typename T>
+	T rangeCheckHelper(const std::string& value)
+	{
+		const auto valueAsInteger = std::stoi(value);
+		if constexpr (std::numeric_limits<T>::max() < std::numeric_limits<int>::max())
+		{
+			if (std::numeric_limits<T>::max() < valueAsInteger)
+			{
+				throw std::out_of_range("rangeCheckHelper argument out of range");
+			}
+		}
+		return static_cast<T>(valueAsInteger);
+	}
 }
 
-/**
- * \fn toUppercase(const std::string& str)
- *
- * Converts a string to uppercase.
- *
- * \param str	Source string.
- *
- * \return	Returns the converted string.
- */
-std::string NAS2D::toUppercase(std::string str)
+namespace NAS2D
 {
-	std::transform(std::begin(str), std::end(str), std::begin(str), [](unsigned char c) noexcept->unsigned char { return static_cast<unsigned char>(::toupper(c)); });
-	return str;
-}
 
-std::vector<std::string> NAS2D::split(std::string str, char delim /*= ','*/)
-{
-	const auto potential_count = 1 + std::count(std::begin(str), std::end(str), delim);
-	NAS2D::StringList result{};
-	result.reserve(potential_count);
+	// template char rangeCheckHelper(const std::string& value);
 
-	std::istringstream ss(str);
+	template<>
+	std::string stringTo(const std::string& value)
+	{
+		return value;
+	}
 
-	std::string curString{};
-	while (std::getline(ss, curString, delim))
+	template<>
+	bool stringTo(const std::string& value)
 	{
-		result.push_back(curString);
+		auto v = toLowercase(value);
+		if(v == "true") return true;
+		if(v == "false") return false;
+		throw std::invalid_argument("Value is not 'true' or 'false' value: " + value);
```

Comment:
I would prefer not using a highly abbreviated name such as `v`.

----

For that matter, I would prefer not including the `toLowercase` as part of this method. It takes choice away from the caller on how strict they want to be. Instead, let the caller decide if they should call `toLowercase` themselves.

This solves the above problem, since the `value` parameter can then be used throughout. There is no need for a secondary modified variable.

Reply:
Normalizing strings is standard practice when comparing them; otherwise you'd have to write:

```cpp
if(value == "true" || value == "TRUE" || value == "True" || value == "tRue" || value == "trUe" || value == "truE") return true;
//...etc
```

In this particular case, taking the argument by value to let the compiler generate a copy is probably better.

Reply:
In this case it would be `stringTo<bool>(toLowercase(value))`. That's easy enough to deal with. It also doesn't take away my choice if I want a case sensitive compare, in which case I can write `stringTo<bool>(value)`.

By trying to be too clever and doing too much, we actually limit expressability and composability.

----

If we get rid of the `toLowercase`, the string use becomes entirely read-only. That removes the potential for move semantic optimizations.